src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -888,70 +1039,224 @@ scaleCreditsForMerge SingleRun{} _ = ScaledCredits 0
-        -- merge was just now finished.
+        -- It is guaranteed that totalSteps' >= spentCredits' /unless/
+        -- the merge was just now finished.
         let totalSteps' = totalSteps + stepsDone
-        let totalCredits' = totalCredits + c
-        -- If an exception happens between the next two writes, then only
-        -- totalCreditsVar will be outdated, which is okay because we will
-        -- resupply credits. It also means we can maintain that @readPrimVar
-        -- totalStepsVar >= readPrimVar totalCreditsVar@, /unless/ the merge was
-        -- just now finished.
-        writePrimVar totalStepsVar $! totalSteps + stepsDone
-        writePrimVar totalCreditsVar $! totalCredits + c
+        let spentCredits' = spentCredits + c
+        -- It is guaranteed that @readPrimVar totalStepsVar >=
+        -- readPrimVar spentCreditsVar@, /unless/ the merge was just now
+        -- finished.
+        writePrimVar totalStepsVar $! totalSteps'
+        writePrimVar spentCreditsVar $! spentCredits'
         assert (case stepResult of
-                  MergeInProgress -> totalSteps' >= totalCredits'
-                  MergeComplete   -> True
-               ) $ pure ()
-
-        pure $ stepResult == MergeComplete
-    when mergeIsDone $
-      modifyMVarMasked_ var $ \case
-        mr@CompletedMerge{} -> pure $! mr
-        (OngoingMerge rs _totalStepsVar _totalCreditsVar m) -> do
-          -- TODO: we'll likely move away from this style of reference counting,
-          -- so this code will change in the future.
-          RefCount n <- Merge.readRefCount m
-          let !n' = fromIntegralChecked n
-          V.forM_ rs $ \r -> Run.removeReferenceN r n'
-          r <- Merge.complete m
-          Merge.removeReferenceN m n'
-          pure $! CompletedMerge r
+                  MergeInProgress -> totalSteps' >= spentCredits'
+                  MergeDone       -> True
+              ) $ pure ()
+
+        pure $ stepResult == MergeDone
+
+{-# SPECIALISE completeMerge ::
+     StrictMVar IO (MergingRunState IO h)
+  -> MutVar RealWorld MergeKnownCompleted
+  -> IO () #-}
+-- | Convert an 'OngoingMerge' to a 'CompletedMerge'.
+completeMerge ::
+     (MonadSTM m, MonadST m, MonadMVar m, MonadMask m, MonadFix m)
+  => StrictMVar m (MergingRunState m h)
+  -> MutVar (PrimState m) MergeKnownCompleted
+  -> m ()
+completeMerge mergeVar mergeKnownCompletedVar = do
+    modifyMVarMasked_ mergeVar $ \case
+      mr@CompletedMerge{} -> pure $! mr
+      (OngoingMerge rs _spentCreditsVar m) -> do
+        -- TODO: we'll likely move away from this style of reference counting,
+        -- so this code will change in the future.
+        RefCount n <- Merge.readRefCount m
+        let !n' = fromIntegralChecked n
+        V.forM_ rs $ \r -> Run.removeReferenceN r n'
+        r <- Merge.complete m
+        Merge.removeReferenceN m n'
+        pure $! CompletedMerge r
+    writeMutVar mergeKnownCompletedVar MergeKnownCompleted
```

Comment:
A weird edge case that can happen: If an async exception occurs before writing `MergeKnownCompleted`, next time around, we will not be able to skip the work in `supplyMergeCredits`. Then `stepMerge` returns False, so we don't enter `completeMerge` and we will forever stay at `MergeMaybeCompleted` and keep calling `stepMerge` on a complete merge.

No real problem, as it's unlikely and the only consequence is some performance overhead, but another reminder of how tricky these things are. We should probably change `stepMerge` to return True on `CompletedMerge` and add a comment about what the returned `Bool` really means (not "it just completed", but "it is complete"). Maybe even return `StepResult` instead, which has more meaning than `Bool`.

Also, if we write to `mergeKnownCompletedVar` in `supplyMergeCredits` instead of here, its usage as a cache is localised to that function and more isolated from the rest of the code. Makes it easier to understand what it does.